File: tests/machine/x/fortran/group_by_sort.f90
program group_by_sort
  implicit none
  type :: Item
    character(len=1) :: cat
    integer :: val
  end type Item
  type(Item) :: items(4)
  integer :: total_a, total_b
  real :: t1, t2

  items(1) = Item('a',3)
  items(2) = Item('a',1)
  items(3) = Item('b',5)
  items(4) = Item('b',2)

  total_a = 0
  total_b = 0
  total_a = items(1)%val + items(2)%val
  total_b = items(3)%val + items(4)%val

  if (total_b > total_a) then
    print *, 'map[cat:b total:', total_b, '] map[cat:a total:', total_a, ']'
  else
    print *, 'map[cat:a total:', total_a, '] map[cat:b total:', total_b, ']'
  end if
end program group_by_sort

File: tests/human/x/pas/group_by_sort.pas
program GroupBySort;

type
  TItem = record
    cat: string;
    val: integer;
  end;

  TRow = record
    cat: string;
    total: integer;
  end;

var
  items: array[1..4] of TItem = (
    (cat:'a'; val:3),
    (cat:'a'; val:1),
    (cat:'b'; val:5),
    (cat:'b'; val:2)
  );
  groups: array[1..2] of TRow;
  count: integer = 0;
  i,j: integer;
  found: boolean;
  tmp: TRow;
begin
  for i := 1 to Length(items) do
  begin
    found := False;
    for j := 1 to count do
      if groups[j].cat = items[i].cat then
      begin
        groups[j].total := groups[j].total + items[i].val;
        found := True;
        Break;
      end;
    if not found then
    begin
      Inc(count);
      groups[count].cat := items[i].cat;
      groups[count].total := items[i].val;
    end;
  end;

  { sort descending by total }
  for i := 1 to count-1 do
    for j := i+1 to count do
      if groups[j].total > groups[i].total then
      begin
        tmp := groups[i];
        groups[i] := groups[j];
        groups[j] := tmp;
      end;

  for i := 1 to count do
    Writeln('cat=', groups[i].cat, ' total=', groups[i].total);
end.
